src/Juvix/Compiler/Nockma/Translation/FromAsm.hs

```diff
@@ -0,0 +1,1048 @@
+              recCase = do
+                pushNat 2
+                mul
+                moveTopFromTo AuxStack ValueStack
+                dec
+                callHelper True (Just (BuiltinFunction BuiltinPow2Go)) 2
+          branch baseCase recCase
+      }
+
+callEnum :: (Enum funId, Members '[Compiler] r) => funId -> Natural -> Sem r ()
+callEnum = callFun . UserFunction . Asm.defaultSymbol . fromIntegral . fromEnum
+
+callFun :: (Members '[Compiler] r) => FunctionId -> Natural -> Sem r ()
+callFun = callHelper False . Just
+
+tcallFun :: (Members '[Compiler] r) => FunctionId -> Natural -> Sem r ()
+tcallFun = callHelper True . Just
+
+getFunctionPath' :: (Members '[Reader CompilerCtx] r) => FunctionId -> Sem r Path
+getFunctionPath' funName = asks (^?! compilerFunctionInfos . at funName . _Just . functionInfoPath)
+
+-- | obj[relPath] := newVal
+-- relPath is relative to obj
+replaceSubterm :: Term Natural -> Path -> Term Natural -> Term Natural
+replaceSubterm obj relPath newVal = OpReplace # (relPath # newVal) # obj
+
+evaluated :: Term Natural -> Term Natural
+evaluated t = OpApply # (OpAddress # emptyPath) # t
+
+-- | The same as replaceSubterm but the path is a cell that is evaluated.
+-- i.e. replaceSubterm a p b = replaceSubterm' a (quote p) b
+replaceSubterm' :: Term Natural -> Term Natural -> Term Natural -> Term Natural
+replaceSubterm' obj relPath newVal =
+  evaluated $ (OpQuote # OpReplace) # ((relPath # (OpQuote # newVal)) # (OpQuote # obj))
+
+sre :: (Members '[Output (Term Natural), Reader CompilerCtx] r) => Sem (Compiler ': r) x -> Sem r x
+sre = subsume . re
+
+-- | funName is Nothing when we call a closure at the top of the stack
+callHelper' ::
+  (Members '[Output (Term Natural), Reader CompilerCtx] r) =>
+  Bool ->
+  Maybe FunctionId ->
+  Natural ->
+  Sem r ()
+callHelper' isTail funName funArgsNum = do
+  let isClosure = isNothing funName
+  -- 1. Obtain the path to the function
+  funPath <- maybe (return (topOfStack ValueStack)) getFunctionPath' funName
+
+  -- 2.
+  --   i. Take a copy of the value stack without the arguments to the function
+  --   ii. Push this copy to the call stack
+  let storeOnStack
+        | isTail = replaceOnStack
+        | otherwise = pushOnStack
+      numToPop
+        | isClosure = 1 + funArgsNum
+        | otherwise = funArgsNum
+
+  unless isTail $ do
+    let frame = makeActivationFrame $ \case
+          ActivationFrameValueStack -> stackPop ValueStack numToPop
+          ActivationFrameTempStack -> OpAddress # stackPath TempStack
+          ActivationFrameAuxStack -> OpAddress # stackPath AuxStack
+    output (pushOnStack FrameStack frame)
+
+  -- 3.
+  --  i. Take a copy of the function from the function library
+  --  ii. Replace its argument area with the arguments from the value stack
+  --  iii. Push this copy to the current function stack
+
+  -- Setup function to call with its arguments
+  -- given n, we compute [R..R] of length n
+  if
```

Comment:
Oh dear, you're emulating a call stack on top of a Nock VM! No wonder it's slow. You should not be handling the call stack explicitly -- Nock VM already does this and you should use it.

A call should just change the Nock subject (transferring the arguments, etc) and evaluate the compiled function call with the new subject. The result of the evaluation should just push to the top of the value stack in the calling function.

The `ret` instruction is then just identity -- returns the top of the value stack. I perhaps should've made it more clear that in JuvixAsm `ret` can only appear in tail positions (like `tcall`).

For example,
```
push arg1
push arg2
call f;
P;
```
should be translated to Nock similar to:
```
[Push arg1*
  [Push arg2*
    [Push
        ["replace subject with new one, with 2 arguments from top of the stack"; fetch & eval f]
        ["remove the 2nd & 3rd value from top of the value stack (the arguments) - that's one indexing and one replace in Nock"; P*]
    ]
  ]
]
```
where `X*` is the translation of `X`, and `Push` is the nock "push" combinator 8.

In fact, almost always the call in JuvixAsm will be like above with the pushes corresponding to the number of arguments. Then we can compile this to slightly better:
```
[Push
  [Push arg1*
    [Push arg2*
      ["replace subject with new one, with 2 arguments from top of the stack"; fetch & eval f]
    ]
  ]
  P*
]
```
(we could in fact always recreate the original call tree and avoid removing the arguments after call return).

It also seems merging the stacks as I initially intended (#2559) might lead to a simpler nock code for "replace subject with new one" (but it's a bit more difficult, because you need to keep track of the value/temporary stack heights).

It matters a lot here whether you generate 1 or 10 instructions for a common operation. This translates to roughly an order of magnitude difference in the running time of generated programs.